src/ir/block_builder.h

```diff
@@ -49,9 +49,11 @@ class BlockBuilder {
   // Adds an operation to the block and returns the operation.
   const Operation& AddOperation(const Operator& op,
                                 NamedAttributeMap attributes,
-                                NamedValueMap inputs) {
-    block_->operations_.push_back(std::make_unique<Operation>(
-        block_.get(), op, std::move(attributes), std::move(inputs)));
+                                NamedValueMap inputs,
+                                std::unique_ptr<Module> impl_module = nullptr) {
+    block_->operations_.push_back(
+        std::make_unique<Operation>(block_.get(), op, std::move(attributes),
+                                    std::move(inputs), std::move(impl_module)));
```

Comment:
I remember previously that I said that I thought the `impl_module` should go on the `Operator` rather than the `Operation`, and you had a reason why it should go on the `Operation` instead, but I don't think we ever fleshed that out together. Why should the implementation live on the `Operation` rather than the `Operator`?

Reply:
Contrary to what the comment associated with `impl_module` suggest, the intention is not to break down an operation into smaller operations. Rather, this is a mechanism to associate lambdas with operations and was something that I pivoted to based on the proposal you had to model operations like `sql.select` and `sql.where` with lambdas.  I left the comment unchanged after the pivot and forgot this context. :( Anyway, having a mechanism to associate lambdas with operations simplifies the modeling of such operations as well other common ones like `if` and `while`.  In fact, it might make sense to have more than one module associated with an operation. e.g,. An `if` operation could have two modules correspnding to the `if` and `else` branches. However,  we can tackle it later.

Reply:
We should probably change the name of the field as well.

Reply:
Interesting, ok. So this is more akin to associating a dynamic call instruction with its possible static targets than associating a function name with its implementation.

It does seem like hanging a `Module` off of an `Operation` could be useful in that circumstance. You could even imagine a kind of attribute called `ResolutionNumberAttribute` that, for some named attribute indicating a resolution to be filled in, associates that attribute with a position in a `vector` of `Module`s. But as you said, that can be tackled later.